Add provider json configuration, adding a dummy Coverr provider

diegovar opened the pull request:
## Summary

Refactors most of the media pane logic to depend on a set of provided json configs. Adds a dummy coverr provider that currently just delegates to Unsplash with a modified query.

<img width="405" alt="Screen Shot 2020-08-05 at 1 24 02 pm" src="https://user-images.githubusercontent.com/512647/89368467-f8b7b080-d71e-11ea-99ab-0efb749e923b.png">

## Relevant Technical Choices

The Coverr "Flag" is in providerConfiguration.js because the providers list is referenced inside hooks, so I can't inject useFeatures to access our normal flag map there (as it would mean using a hook inside another hook).

## To-do

Could extract the per-provider error messages into the configuration as well.

## User-facing changes

Adds a new provider, makes them configurable.

## Testing Instructions

Enable media3pTab` flag and set `showCoverrTab` in `providerConfiguration.js` to true.

---

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #2804
Fixes #3349
Fixes #3355

diegovar committed:
First working version of coverr integration

diegovar committed:
use a previousProps hack to not reload media when switching tabs

diegovar committed:
Ensure we don't request categories if they've already been fetched.

diegovar committed:
fix test

diegovar committed:
use ref hack to avoid relying on previous props.

diegovar committed:
Revert "use ref hack to avoid relying on previous props."

This reverts commit fe96d665edb784d91dce705778269d851efa99f7.

diegovar committed:
move showSnackbar

diegovar committed:
Add an internal "flag" for the coverr subtab.

diegovar committed:
Merge branch 'main' of https://github.com/google/web-stories-wp into provider-json

diegovar committed:
Merge branch 'main' of https://github.com/google/web-stories-wp into provider-json

diegovar committed:
move attribution styles

diegovar committed:
remove comment

cvolzke4 commented on assets/src/edit-story/app/media/media3p/providerConfiguration.js:
```diff
@@ -0,0 +1,132 @@
+          supportedContentTypes: [ContentType.IMAGE],
+          supportsCategories: true,
+          requiresAuthorAttribution: true,
+          attributionBuilder: UnsplashAttribution,
```
[nit]: Since this is a component, should this be named attributionComponent or attributionPill?

diegovar replied:
Done

cvolzke4 commented on assets/src/edit-story/app/media/media3p/api/media3pApiProvider.js:
```diff
@@ -67,9 +59,20 @@ function Media3pApiProvider({ children }) {
   }) {
-    if (provider.toLowerCase() !== Providers.UNSPLASH) {
+    if (
+      provider !== ProviderType.UNSPLASH &&
```
Perhaps this could be:
```
if (!Object.keys(Providers).contains(provider)) {
  throw ...;
}
```

diegovar replied:
Done.

cvolzke4 commented on assets/src/edit-story/app/media/media3p/providerConfiguration.js:
```diff
@@ -0,0 +1,132 @@
+  );
+}
+
+const showCoverrTab = false;
```
To keep the features in useFeatures(), perhaps the name of the flag can be added as an attribute?
```
featureName: 'showCoverrTab'
```

cvolzke4 replied:
The config can then be specified as a simple object:
```
export const providers = {
  [ProviderType.UNSPLASH]: {
    displayName: 'Unsplash',
    supportedContentTypes: [ContentType.IMAGE],
    supportsCategories: true,
    requiresAuthorAttribution: true,
    attributionBuilder: UnsplashAttribution,
  },
  [ProviderType.COVERR]: {
    displayName: 'Coverr',
    supportedContentTypes: [ContentType.VIDEO],
    supportsCategories: false,
    requiresAuthorAttribution: false,
    attributionBuilder: CoverrAttribution,
    featureName: 'showCoverrTab',
  },
}
```

diegovar replied:
Good idea!

cvolzke4 commented on assets/src/edit-story/app/media/media3p/test/useFetchMediaEffect.js:
```diff
@@ -60,7 +60,7 @@ describe('useFetchMediaEffect', () => {
           provider: 'unsplash',
           selectedProvider: 'unsplash',
-          pageToken: 'pageToken',
+          pageToken: undefined,
```
Why was this part changed? This was testing that the current page's token is propagated through fetchMediaStart, fetchMediaSuccess etc.

diegovar replied:
So with this current implementation, the media is only fetched if no media has been fetched (pageToken and nextPageToken are both falsey) or sth has changed from the previous render. I modified the test to ensure that the fetch would happen, and I wasn't able to figure out how to trigger the case where dependencies of a useEffect hook change. Do you know how to do that?

cvolzke4 replied:
Did you try calling renderUseFetchMediaEffect() again with a different property value for one of the dependencies? I think that will re-trigger the useEffect.

diegovar replied:
I did, it re-triggers the useEffect but the previous props are not kept at all. I suspect because it's a completely different instance.

cvolzke4 replied:
Interesting, perhaps useRef is only maintained within the same act() call?

diegovar replied:
yeah ,tried running the useFetchMediaEffect in the same act, and even inside the same renderHook, but the useRef is not maintained :(

cvolzke4 commented on assets/src/edit-story/app/media/media3p/providerConfiguration.js:
```diff
@@ -0,0 +1,132 @@
+ *
+ * @type {Object.<string, ProviderConfiguration>}
+ */
+export const Providers = Object.fromEntries(
```
[nit]: 'Providers' returns a variable not a class, so it should have a lowercase 'p'. Or perhaps it makes sense to be the default export?

diegovar replied:
I think it's the same as providerType.js right? Not sure why it would be lowercase p?

cvolzke4 replied:
ProviderType is an enum, whereas this is now a field:
go/js-style#naming-enum-names
go/js-style#naming-non-constant-field-names

diegovar replied:
Given this is a constant name, I've renamed it to PROVIDERS.

cvolzke4 commented on assets/src/edit-story/app/media/media3p/providerConfiguration.js:
```diff
@@ -0,0 +1,132 @@
+ * React Component for this provider.
+ */
+
+function UnsplashAttribution() {
```
This file is rendering AttributionPill and providing providerConfiguration, perhaps the AttributionPill parts belong in a separate file?

diegovar replied:
Done.

cvolzke4 commented on assets/src/edit-story/app/media/media3p/providerReducer.js:
```diff
@@ -17,7 +17,7 @@
  * Internal dependencies
  */
-import commonReducer, {
+import paginationReducer, {
```
Thanks :)

cvolzke4 commented on assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js:
```diff
@@ -27,12 +27,15 @@ import { useEffect } from 'react';
   if (provider === ProviderType.UNSPLASH) {
     return __('Error loading media from Unsplash', 'web-stories');
+  } else if (provider === ProviderType.COVERR) {
+    return __('Error loading media from Coverr', 'web-stories');
```
The display names ('Unsplash' or 'Coverr') should be read from the provider configuration.

diegovar replied:
I couldn't find a way of doing i18n message interpolation using __ (seems very basic), so I've moved the messages to the config.

winterDroid commented on assets/src/edit-story/app/media/media3p/providerConfiguration.js:
```diff
@@ -0,0 +1,132 @@
+              supportedContentTypes: [ContentType.VIDEO],
+              supportsCategories: false,
+              requiresAuthorAttribution: false,
+              attributionBuilder: CoverrAttribution,
```
Why did you not just put the properties for the attribution here instead of the whole component?

diegovar replied:
Because we might need different styles of attribution per provider. Maybe Unsplash is ok with a "Powered by <SVG>" but Tenor wants something different.

winterDroid replied:
That makes sense. But we can define what needs to be customized and provided. It does not just need to be one image or so.

diegovar replied:
I think I'll leave this as is, and in the meantime will confirm with Tenor and Unsplash if this is the attribution format they both expect.

winterDroid commented on assets/src/edit-story/app/media/media3p/providerConfiguration.js:
```diff
@@ -0,0 +1,132 @@
+  new Map(
+    [
+      [
+        ProviderType.UNSPLASH,
```
I think we should remove the whole `ProviderType` construct and instead just define a unique provider key in here. 

Everywhere where we currently use a check whether the provider matches a certain `ProviderType` should be avoided and be part of the configuration.

diegovar replied:
I'd rather do that in a separate PR, that might be quite a bit of work and this PR is big enough.

winterDroid replied:
Yeah that's fair 👍

winterDroid replied:
Do you want to create a follow-up ticket for it or do it in the same? If latter one, this probably should just address and not fix it according to the summary.

diegovar replied:
I've created #3707

winterDroid commented on assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:
```diff
@@ -77,49 +97,47 @@ function Media3pPane(props) {
-    if (isActive) {
-      setSelectedProvider({ provider: 'unsplash' });
+    if (isActive && !selectedProvider) {
+      setSelectedProvider({ provider: ProviderType.UNSPLASH });
```
Should it be whatever is first in the providers config?

diegovar replied:
Good catch.

cvolzke4 commented on assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:
```diff
@@ -77,49 +97,47 @@ function Media3pPane(props) {
   useEffect(() => {
-    if (isActive) {
-      setSelectedProvider({ provider: 'unsplash' });
+    if (isActive && !selectedProvider) {
```
What's the reason for the !selectedProvider check?

cvolzke4 replied:
Nevermind, I understand it now :)

winterDroid commented on assets/src/edit-story/app/media/media3p/providerConfiguration.js:
```diff
@@ -0,0 +1,132 @@
+        },
+      ],
+      showCoverrTab
+        ? [
```
nit: can't this be `showCoverrTab && [...]`

diegovar replied:
I've moved to use featureName so we can have it as a feature flag.

cvolzke4 commented on assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:
```diff
@@ -130,32 +148,65 @@ function Media3pPane(props) {
+                {Providers[providerType].supportsCategories && (
+                  <Media3pCategories
+                    categories={
+                      media3p[providerType].state.categories.categories
```
Could media3p[providerType].state / edia3p[providerType].actions be extracted to local variables?
```
<Media3pCategories
    categories={providerState.categories.categories}
    selectedCategoryId={providerState.categories.selectedCategoryId}
    selectCategory={providerActions.selectCategory}
    deselectCategory={providerActions.deselectCategory}
```

diegovar replied:
Done.

cvolzke4 commented on assets/src/edit-story/components/library/panes/media/common/paginatedMediaGallery.js:
```diff
@@ -197,20 +172,7 @@ function PaginatedMediaGallery({
-          </AttributionPill>
-        </a>
-      )}
+      {Providers[providerType].attributionBuilder()}
```
To make it clear what 'attributionBuilder' is, I would suggest:
```
  const AttributionComponent = Providers[providerType].attributionBuilder;
  return (
    <>
      <MediaGalleryContainer
        data-testid="media-gallery-container"
        ref={refCallbackContainer}
      >
        <MediaGalleryInnerContainer>{mediaGallery}</MediaGalleryInnerContainer>
      </MediaGalleryContainer>
      <AttributionComponent />
    </>
  );
```

diegovar replied:
I renamed the field to attributionComponent so it should be clear now.

cvolzke4 commented on assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js:
```diff
@@ -53,41 +57,66 @@ export default function useFetchMediaEffect({
+        searchTerm != previousProps.searchTerm ||
+        selectedCategoryId != previousProps.selectedCategoryId);
+    const firstFetchOrSomethingChanged =
+      (!pageToken && !nextPageToken) || somethingChanged;
```
I don't think you need 'nextPageToken' to know if it's the first fetch. pageToken will be undefined if the first page is being fetched.

diegovar replied:
I think if you've already fetched data, nextPageToken is not null but pageToken is. We don't want to fetch if we've already fetched in that case, so I need to check both?

cvolzke4 replied:
ahh, you are right.
I wonder if that's a problem, ie. if there is <= 1 page of images, then pageToken and nextPageToken will both be undefined, even if the first fetch has already completed.

diegovar replied:
Good point, I've modified the logic to depend on isMediaLoading and isMediaLoaded instead. If both are false, that means that no media has been fetched yet.

cvolzke4 commented on assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js:
```diff
@@ -53,41 +57,66 @@ export default function useFetchMediaEffect({
       fetch();
     }
+    // We don't want to depend on previousProps, see https://blog.logrocket.com/how-to-get-previous-props-state-with-react-hooks/
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Another way to avoid depending on previousProps, you could use the previousProps reference inside of the useEffect block:

```
  const previousPropsRef = useRef();
  useEffect(() => {
    const previousProps = previousPropsRef.current;
    previousPropsRef.current = { pageToken, searchTerm, selectedCategoryId };
    // ...
  }, [
    // The ref never changes
    previousPropsRef
  ]);
```

diegovar replied:
I'll try this out and report back.

diegovar replied:
I don't think this would do what we want. Accessing previousPropsRef.current inside the effect would not give you the previous value of the props as it's updated inside an effect in usePrevious.

cvolzke4 replied:
The usePrevious hook would not be used in the example given above.
previousPropsRef.current is updated in the useFetchMediaEffect's useEffect().

diegovar committed:
Merge branch 'main' of https://github.com/google/web-stories-wp into provider-json

diegovar committed:
First round of review feedback

diegovar committed:
Merge branch 'main' of https://github.com/google/web-stories-wp into provider-json

diegovar committed:
fix includes, use isMedia{Loading|Loaded} to detect first render

diegovar committed:
Disable showCoverrTab

winterDroid commented on assets/src/edit-story/app/media/media3p/api/media3pApiProvider.js:
```diff
@@ -67,9 +60,17 @@ function Media3pApiProvider({ children }) {
+
+    // Temporary hack alert!: Convert coverr to unsplash for testing until
+    // Coverr backend is implemented.
+    if (provider === ProviderType.COVERR) {
```
Refer to a ticket.

diegovar replied:
Done

winterDroid commented on assets/src/edit-story/app/media/media3p/attribution.js:
```diff
@@ -0,0 +1,77 @@
+  flex-wrap: nowrap;
+  font-size: 12px;
+  color: ${theme.colors.fg.white};
+  background-color: ${rgba(theme.colors.bg.black, 0.7)};
```
In most places we use `${({ theme }) => theme.fonts.duration.family}` and then don't import the `theme` directly

diegovar replied:
Done.

cvolzke4 commented on assets/src/edit-story/app/media/media3p/useFetchCategoriesEffect.js:
```diff
@@ -54,21 +47,30 @@ export default function useFetchCategoriesEffect({
+          provider,
+        });
+        fetchCategoriesSuccess({ provider, categories: newCategories });
+      } catch (e) {
```
The 'e' is unused here.

winterDroid commented on assets/src/edit-story/app/media/usePrevious.js:
```diff
@@ -0,0 +1,34 @@
+ * @param {Object} value Object to keep track of.
+ * @return {Object} The previous value of the object.
+ */
+export default function usePrevious(value) {
```
Shall we have a test for that?

diegovar replied:
Removes this file.

winterDroid commented on assets/src/edit-story/components/library/panes/media/common/paginatedMediaGallery.js:
```diff
@@ -197,20 +172,7 @@ function PaginatedMediaGallery({
-          </AttributionPill>
-        </a>
-      )}
+      {PROVIDERS[providerType].attributionComponent()}
```
I think calling it a builder did make more sense as it build the component.

cvolzke4 reviewed (approved):
Looking good!

diegovar committed:
Merge branch 'main' of https://github.com/google/web-stories-wp into provider-json

diegovar committed:
attribution, loading pill fixes, remove usePRevious

- don't show attribution for local
- ensure the loading pill is shown correctly in some intermediate cases
- use better approach instead of usePrevious

diegovar committed:
Remove usePrevious

diegovar committed:
review feedback

diegovar committed:
better css

diegovar committed:
Merge branch 'main' of https://github.com/google/web-stories-wp into provider-json

diegovar committed:
re-add media3pTab flag, reorder experiments
